### PR TITLE
Avoid triggering type-check assert for cyfuncs [0.29.x]

### LINKED ISSUE #4804: [BUG] Assertion failure in PyCFunction_GET_FLAGS on debug mode
**Describe the bug**
When CPython is compiled in debug mode, Cython may trigger an assertion related to function flags.

**To Reproduce**

* Compile CPython 3.11.0b1 from source `--with-pydebug`
* `pip install asynq` with the newly compiled python
* `python -c 'import asynq'`
* Crash: `Assertion failed: (PyCFunction_Check(func)), function PyCFunction_GET_FLAGS, file methodobject.h, line 59.`

**Expected behavior**
No crash

**Environment (please complete the following information):**
 - OS: macOS
 - Python version 3.11b1
 - Cython version 0.29.30

**Additional context**

The crash is around here: https://github.com/cython/cython/blob/169876872f3cb6198971a1db07e5b8a9d12b3dac/Cython/Utility/ObjectHandling.c#L2237. If `func` is a `CyFunctionObject`, the `__Pyx_IsCyOrPyCFunction` check passes and we call `PyCFunction_GET_FLAGS`, but in debug mode that function checks that its arg is exactly a PyCFunction.

It appears that this is harmless because the layout of `CyFunctionObject` has a `PyCFunction` at the top, so it's fine at runtime to treat the former as if it's the latter. Therefore, things work fine in non-debug mode.

### PR DESCRIPTION
**Pull request by da-woods**
Fixes #4804 (but not hugely satisfactorally)

### CODE REVIEW

File: Cython/Utility/ObjectHandling.c [hunk 1]
```diff
@@ -2322,7 +2322,10 @@ static CYTHON_INLINE PyObject* __Pyx_PyObject_CallNoArg(PyObject *func) {
         return __Pyx_PyFunction_FastCall(func, NULL, 0);
     }
 #endif
-#ifdef __Pyx_CyFunction_USED
+#ifdef __Pyx_CyFunction_USED && defined(NDEBUG)
```
Comment (stefanor) [inline Cython/Utility/ObjectHandling.c:2325]:
```suggestion
#if defined(__Pyx_CyFunction_USED) && defined(NDEBUG)
```

Comment (da-woods) [reply Cython/Utility/ObjectHandling.c]: Thanks - not sure how I missed that or why it didn't break

Comment (stefanor) [reply Cython/Utility/ObjectHandling.c]: I think gcc just ignored everything after the first term (and printed a warning)

**Comment by scoder** [comment]
Looks ok to me. It's a pure optimisation, we can live without that if it gets in the way.

**Pull request merged**